### ISSUE TITLE
cached_attachment_data is funky when validations fail

### ISSUE THREAD

**Issue by ivanyv** [janko-m/shrine#65]
There are two scenarios:

1) Choose a file that passes validations but the record itself has other validation errors. The file selected is stuck, can't upload a different file (upon submit the previous file is still assigned)

2) Add a file that fails validations. Same result. Can't even change it for a valid file.

If I remove the `cached_logo_data` attribute from the strong parameters filter, then the file is no longer stuck (one can change it), but if the record has validation errors, upon fixing those and resubmitting (without changing the already valid file), the file is then not sent.

I'm banging my head trying some conditionals while manually setting the `cached_attachment_attribute` but can't figure out what causes this yet :(

**Comment by janko-m** [comment]
I've managed to reproduce the issue. It is not specific to `f.hidden_field :cached_logo_data`, it also happens when you use regular `f.hidden_field :logo`. I will push a fix shortly.

**Comment by ivanyv** [comment]
Thank god, I was starting to switch to CarrierWave (:cringe:) -- Thanks!

**Comment by janko-m** [comment]
Sorry, had to go AFK for a while.

When you use `hidden_field :logo`, then everything will always work properly, because the `file_field :logo` will always overwrite the hidden field, since the fields are named the same so browser will always choose the field in the form (which is the file field).

At the moment you can just not use cached_attachment_data:

```erb
<%= form_for @company do |f| %>
  <%= f.hidden_field :logo, value: @company.logo_data
  <%= f.file_field :logo
  <!-- ... -->
<% end %>
```

Even though cached_attachment_data's behaviour is to set value to `@company.logo_data if @company.logo_attacher.cached?`, I now think it makes more sense without the `if` statement, because the "file attribute" always being sent is consistent with other normal fields in the form.

The reason why cached_attachment_data doesn't work the same is because fields are named differently, so the hidden field will always be sent regardless if the file field was set. I will think more if it's worth fixing, but I will probably deprecate this plugin over the default solution above. Sorry for the troubles!

**Comment by janko-m** [comment]
Ok, I think I've figured out exactly what happened. I bet you were permitting your parameters like this:

```rb
params.require(:company).permit(:logo, :cached_logo_data)
```

What happens here is that, even though the original params hash was sent with keys in order of `cached_logo_data, logo`, this permit call *reverses* the order to `logo, cached_logo_data`. The order in the hash is the order in which Shrine setters are called, and Shrine relies on the original order. Now first your new file is assigned, and then the cached (invalid) image, overwriting the new file.

I don't use Rails anymore, so I totally didn't predict that you can naturally reverse the order of parameters with strong_params. I will certainly deprecate `#cached_<attachment>_data=` so that you cannot use it in forms, but I will keep the getter, and this will now be the recommended way (so please use it):

```rb
<%= form_for @company do |f| %>
  <%= f.hidden_field :logo, value: @company.cached_logo_data
  <%= f.file_field :logo
  <!-- ... -->
<% end %>
```

**Comment by ivanyv** [comment]
I did actually ponder that for a second but I didn't think it made sense and then forgot to test it.

So I tried your previous solution and that does work; what would the difference be between assigning the hidden field's value to `cached_logo_data` instead of `logo_data`?

**Comment by janko-m** [comment]
There is currently no difference in the end behaviour. However, while rethinking about the whole process of attaching, I've found one subtle bug that could happen with backgrounding:

If the user attaches the file and saves the form, the record will be first saved with the cached attachment (and a background job is kicked of which does the processing of the cached file and uploading to permanent storage). However, if the user would revisit the form before the background job has finished, which means the hidden field will be populated with the cached attachment, if the user submits the form but the background job happens to finish before that, the stored file would again get overwritten with the cached file. This shouldn't cause any errors, the only side effect would be that the background job is repeated, but I would still like to fix this.

So I will now make `cached_logo_data` return the cached file only if it was changed, which practically means that the hidden field will be populated only in case of validation errors, not on normal form submission when the attachment is already saved anyway. I think that's the correct behaviour.

Attaching logic is hard 😁
